dask/array/slicing.py

```diff
@@ -697,6 +693,72 @@ def replace_ellipsis(n, index):
             index[loc + 1:])
 
 
+def normalize_slice(idx, dim):
+    """ Normalize slices to canonical form
+
+    Parameters
+    ----------
+    idx: slice or other index
+    dim: dimension length
+
+    Examples
+    --------
+    >>> normalize_slice(slice(0, 10, 1), 10)
+    slice(None, None, None)
+    """
+
+    if isinstance(idx, slice):
+        start, stop, step = idx.start, idx.stop, idx.step
+        if start is not None:
+            while start < 0:
+                start += dim
+        if stop is not None:
+            while stop < 0:
+                stop += dim
+        if start == 0:
+            start = None
+        if stop == dim:
+            stop = None
+        if step == 1:
+            step = None
+        return slice(start, stop, step)
+    return idx
+
+
+def normalize_index(idx, shape):
+    """ Normalize slicing indexes
+
+    1.  Replaces ellipses with many full slices
+    2.  Adds full slices to end of index
+    3.  Checks bounding conditions
+    4.  Replaces numpy arrays with lists
+    5.  Posify's integers and lists
+    6.  Normalizes slices to canonical form
+    """
```

Comment:
Much of this normalization is repeated in the actual slicing code (e.g. `replace_ellipsis` is called in `slice_array`). Should probably clean up the repeated work?

Reply:
The challenge here is that all of the current accuracy checks are on the internal functions, not on the public API, so we'll have to rewrite a fair amount of the test suite.  I'll look into how expensive this will be.

Reply:
I've removed checks elsewhere and moved testing logic outside